core/cap-0047.md

```diff
@@ -186,93 +183,67 @@ index 23918226d..3ed9625bc 100644
+`ENVELOPE_TYPE_CONTRACT_ID_FROM_KEY`, which contains a public key, and a salt
+provided by the user. The `create_contract` host function also does some
+validation to ensure the contract and contractID are created securely. It uses
+the `check` functions and `Identifier` defined in CAP-0052 to verify the
+identity of the contract creator. 
+
+The transaction will fail if a contract with the calculated contractID already
+exists.
 
 TODO: Reserve requirement for `ContractDataEntry`?
+TODO: Another envelope type for contractIDs from accountIDs.
+
+```rust
+// Define CREATE_CONTRACT_SEPARATOR =
+//     sha256("create_contract(nonce: u256, contract: Vec<u8>, salt: u256, key: u256, creator: Identifier, sig: Vec<u8>)"),
+// Calls check with
+//     nonce = nonce
+//     params = vec![CREATE_SEPARATOR, salt, creator, sha256(contract)]
+//     key = key
+//     sig = sig
+// Creates a new contract.
+fn create_contract(nonce: u256, contract: Vec<u8>, salt: Vec<u8>, key: u256, creator: Identifier, sig: Vec<u8>) -> ContractID;
+
+// Same as above but replaces the single key and sig with an AccountID
+// and a list of signatures. For account based authorization.
+fn create_contract(nonce: u256, contract: Vec<u8>, salt: Vec<u8>, key: AccountID, creator: Identifier, sig: Vec<(u256, Vec<u8>)>)) -> ContractID;
+```
 
-### Host function to create contracts
+### Creating a contract within a contract
 Factory contracts are quite popular already on other networks, so this CAP adds
-functionality to support them. The contractID for these contracts will be
-generated a little differently than a contract created by an account. The
-`add_contract` call will create the new contractID by taking the SHA-256 hash of
-a `HashIDPreimage`, with type `ENVELOPE_TYPE_CHILD_CONTRACT_ID`, which contains
-the parent contractID, and a salt that the parent contract passes to
-`create_contract`.
+functionality to support them. The contractID for these contracts can be
+generated a little differently than a contract created using
+`InvokeHostFunctionOp` for simplicity. The `create_contract_using_contractID`
+call will create the new contractID by taking the SHA-256 hash of a
+`HashIDPreimage`, with type `ENVELOPE_TYPE_CONTRACT_ID_FROM_CONTRACT`, which
+contains the parent contractID, and a salt that the parent contract passes to
+`create_contract_using_contractID`.
+
+`create_contract_using_contractID` will trap if it is not called from within
+another contract.
+
+Contract creators still have the option to use `create_contract` if they choose to.
+This can be helpful in factory contracts where the salt is specified by the
+contract caller.
 
 ```rust
-// creates a new contract
-fn create_contract(contract: Vec<u8>, salt: Vec<u8>) -> /*Returns new contractID if contract was created. Empty result otherwise*/;
+// Can only be used inside another contract. Will use the parents contractID.
+fn create_contract_using_parent_id(contract: Vec<u8>, salt: Vec<u8>) -> ContractID
```

Comment:
Nitpick: this name doesn't match the name you're using above.

Reply:
fixed